renderer.md

```diff
@@ -0,0 +1,28 @@
+## Disclaimer
+
+First of all this is an attempt to gather information about state of Renderer in dlang
+
+# What is a Renderer?
+ 
+Renderer contains a rendering state and provides functionality to draw primitives using specific color or texture. Renderer uses Image as a buffer to render to. Renderer is immediate one, i.e. it instantly modifies underlying Image during calling its methods. Renderer draw the following primitives:
+- point
+- line
+- polyline
+- polygone (probably convex only?)
+- rectangle
+- ellipse
+- text
+- image
+
+Lines can be stroked, 2D figures can be stroked and filled. Rectange can have rounded corners.
+
+Example of renderer: 
+- [nanovega](http://dpldocs.info/experimental-docs/arsd.nanovega.html)
+- [SDL_Renderer](https://wiki.libsdl.org/SDL_Renderer)
+- [cairo](https://www.cairographics.org)
+  
+  
+  # Open questions
+   - What dimensionality should it be? Probably 2D only?
+   - Is it possible and convinient to have both compile-time and runtime API?
+   - what API should it have?
```

Comment:
The only successful low-level UI rendering API I'm aware of is this: https://github.com/ocornut/imgui/blob/master/imgui.h#L1781

Something like that with a D flavour would make a good foundation.

Reply:
Not sure why it would have a compile-time interface, I mean I can see it templated by the color type maybe, but not necessarily either.

Reply:
There is [dimgui](https://github.com/d-gamedev-team/dimgui) (not very active). Also there is [nuklear](https://github.com/vurtun/nuklear). I used both `dimgui`, `dearimgui` and `nuklear`. None of them satisfy me but `nuklear` is more advanced. The drawback it has is poor layouting. But I'd like to make a little research what could be done with it.

Reply:
Also [nanovega](https://github.com/adamdruppe/arsd/blob/master/nanovega.d) should be mentioned in this context

Reply:
We write our editors with dearimgui. It layouts just fine.
There's nothing incompatible with decent layout widgets in the library core, you just have to write the layout tools you want. We have done work to support multiple windows, docking, etc.
dearimgui is actually an excellent example of simplicity-first design at the core, and a remarkable extensibility story. If there's a single library I would refer to here, that's the one.

Reply:
What makes nuklear good?

Reply:
I can't give clean answer. I have one single use case - drawing huge listview of treeview and IIRC with dearimgui I failed to navigate through this listview using scrollbar - when list is large it is impossible to scroll to specific item because huge leaps. nuklear lets more fine grained control. But [it lacks](https://github.com/vurtun/nuklear/issues/504) of layouting engine that enable listview to have dynamic height items.